pkg/planner/core/logical_plan_builder.go

```diff
@@ -5718,13 +5721,109 @@ func buildColumns2Handle(
 				return nil, err
 			}
 			end := offset + tblLen
-			cols2Handles = append(cols2Handles, TblColPosInfo{tblID, offset, end, handleCol})
+			cols2Handles = append(cols2Handles, TblColPosInfo{TblID: tblID, Start: offset, End: end, HandleCols: handleCol})
 		}
 	}
 	sort.Sort(cols2Handles)
 	return cols2Handles, nil
 }
 
+// buildColPositionInfoForDelete builds columns to handle mapping for delete.
+// We'll have two kinds of columns seen by DELETE:
+//  1. The columns that are public. They are the columns that not affected by any DDL.
+//  2. The columns that are not public. They are the columns that are affected by DDL.
+//     But we need them because the non-public indexes may rely on them.
+//
+// The two kind of columns forms the whole columns of the table. Public part first.
+// This function records the following things:
+//  1. The position of the columns used by indexes in the DELETE's select's output.
+//  2. The row id's position in the output.
+//  3. The possible reference column of the column under MODIFY/CHANGE ddl job.
+func buildColPositionInfoForDelete(
+	names []*types.FieldName,
+	tblID2Handle map[int64][]util.HandleCols,
+	tblID2Table map[int64]table.Table,
+) (TblColPosInfoSlice, error) {
+	var cols2PosInfos TblColPosInfoSlice
+	for tblID, handleCols := range tblID2Handle {
+		tbl := tblID2Table[tblID]
+		pubCols := tbl.Cols()
+		deletableIdxs := tbl.DeletableIndices()
+		deletableCols := tbl.DeletableCols()
+		tblInfo := tbl.Meta()
+
+		for _, handleCol := range handleCols {
+			curColPosInfo, err := buildSingleTableColPosInfoForDelete(names, handleCol, pubCols, deletableIdxs, deletableCols, tblInfo)
+			if err != nil {
+				return nil, err
+			}
+			cols2PosInfos = append(cols2PosInfos, curColPosInfo)
+		}
+	}
+	sort.Sort(cols2PosInfos)
+	return cols2PosInfos, nil
+}
+
+// buildSingleTableColPosInfoForDelete builds columns mapping for delete.
+func buildSingleTableColPosInfoForDelete(
```

Comment:
Better to add unit test cases to cover the two new methods `buildColPositionInfoForDelete` and `buildSingleTableColPosInfoForDelete`.

Reply:
I'll add some in the next pr that supports the column pruning.